lib/cretonne/src/licm.rs

```diff
@@ -15,40 +20,137 @@ pub fn do_licm(func: &mut Function,
                domtree: &mut DominatorTree,
                loop_analysis: &mut LoopAnalysis) {
     loop_analysis.compute(func, cfg, domtree);
-    for lp in loop_analysis.loops() {
-        // For each loop that we want to optimize we determine the set of loop-invariant
-        // instructions
-        let invariant_inst = remove_loop_invariant_instructions(lp, func, cfg, loop_analysis);
-        // Then we create the loop's pre-header and fill it with the invariant instructions
-        // Then we remove the invariant instructions from the loop body
-        if !invariant_inst.is_empty() {
-            // If the loop has a natural pre-header we use it, otherwise we create it.
-            let mut pos;
-            match has_pre_header(&func.layout, cfg, domtree, loop_analysis.loop_header(lp)) {
-                None => {
-                    let pre_header =
-                        create_pre_header(loop_analysis.loop_header(lp), func, cfg, domtree);
-                    pos = Cursor::new(&mut func.layout);
-                    pos.goto_bottom(pre_header);
-                    pos.prev_inst();
+    let mut inst_pool = ListPool::new();
+    let mut insts_rpo = EntityList::new();
+    // Because Rust doesn't allow us to modify data structures that we are iterating upon, we
+    // first have to do a pass to list all the instruction susceptible to be hoisted.
+    for ebb in domtree.cfg_postorder().iter().rev() {
+        for inst in func.layout.ebb_insts(*ebb) {
+            match loop_analysis.base_loop_inst(inst, &func.layout) {
+                None => (),
+                Some(_) => {
+                    if func.dfg.has_results(inst) {
```

Comment:
I just found the `cretonne::ir::instructions::Opcode::other_side_effects` that should do the job.